STELLAR-1021 Add our own Waterfall SInk

reiinakano opened the pull request:
This change adds our own Waterfall Sink heavily based on the sink from Satnogs. All you really need to review is "waterfall_plotter_impl.cc". Some differences:

1. The `waterfall_plotter` sub-block is now written in C++. This lets us avoid continuous `np.concatenate` calls that may be inefficient. This is now replaced by storing a linked list of pointers to the buffers passed in to `work`. When the flowgraph is stopped, we combine the saved buffers into one and pass it to a `matplotlib` function for generating the .png image.
2. More importantly, we can choose whether to save the .png in the filesystem as before and/or save the .png bytes in memory so the flow graph caller can extract them after the flow graph is closed. This will let Starcoder get the .png bytes from the block through a public function (not yet implemented) and pass it back through gRPC.

I kept the license headers for Satnogs in the `waterfall_heatmap` and `waterfall_sink` hierarchical block since we mostly just copied those.

anuraaga commented on gr-starcoder/python/plot_waterfall.py:
```diff
@@ -0,0 +1,43 @@
+    fig.axes.get_yaxis().set_visible(False)
+    plt.axis('off')
+
+    buf = io.BytesIO()
```
You should pass in `initial_bytes` (I guess `fig` must provide some way of knowing it, otherwise the calculation is of course 8 * 4 * width * height for RGBA)

anuraaga replied:
There's  also a way to convert matplotlib into a numpy array more directly (I guess just two copies, while this has an unknown number of copies due to growing arraay)

anuraaga commented on gr-starcoder/python/plot_waterfall.py:
```diff
@@ -0,0 +1,43 @@
+    plt.axis('off')
+
+    buf = io.BytesIO()
+    plt.savefig(buf, bbox_inches='tight', pad_inches=0)
```
Can we explicitly set the format? Not sure what format this saves as

anuraaga commented on gr-starcoder/python/plot_waterfall.py:
```diff
@@ -0,0 +1,43 @@
+
+    buf = io.BytesIO()
+    plt.savefig(buf, bbox_inches='tight', pad_inches=0)
+    if filename is not None:
```
Let's do this before the above `savefig` since no need to split the two pieces of logic

anuraaga commented on gr-starcoder/python/qa_waterfall_heatmap.py:
```diff
@@ -0,0 +1,41 @@
+from gnuradio import blocks
+import starcoder_swig as starcoder
+
+class qa_waterfall_heatmap (gr_unittest.TestCase):
```
Let's write tests

anuraaga commented on gr-starcoder/python/qa_waterfall_plotter.py:
```diff
@@ -0,0 +1,47 @@
+        op = starcoder.waterfall_plotter(self.fft_size, "")
+        self.tb.connect(src, s2v, op)
+        self.tb.run()
+        # check data
```
Let's do this

anuraaga commented on gr-starcoder/lib/waterfall_plotter_impl.cc:
```diff
@@ -0,0 +1,177 @@
+        static_cast<long int>(d_total_size / (d_fft_size * sizeof(int8_t))),
+        static_cast<long int>(d_fft_size)};
+      const int ND = 2;
+      std::cout << "dims: " << dims[0] << " " << dims[1] << std::endl;
```
Is it normal to have printfs in gnuradio block src?

reiinakano replied:
Some blocks do, some blocks only print when some setting is verbose, and some have printfs commented out. Will just remove them from this, I'm just using them for debug anyway.

anuraaga commented on gr-starcoder/lib/waterfall_plotter_impl.h:
```diff
@@ -0,0 +1,60 @@
+    class waterfall_plotter_impl : public waterfall_plotter
+    {
+     private:
+      std::list<item> d_list_of_arrays;
```
What does `d_` stand for? I've only ever seen the `m_` prefix

reiinakano replied:
Dunno. But it's here under naming conventions https://wiki.gnuradio.org/index.php/Coding_guide_impl

anuraaga commented on gr-starcoder/lib/waterfall_plotter_impl.cc:
```diff
@@ -0,0 +1,177 @@
+              gr::io_signature::make(1, 1, fft_size * sizeof(int8_t)),
+              gr::io_signature::make(0, 0, 0)),
+        d_total_size(0),
+        d_blob(new char[0]),
```
Not sure if this leaks an empty array or not (which would be leaking the array header with no content)

anuraaga commented on gr-starcoder/lib/waterfall_plotter_impl.cc:
```diff
@@ -0,0 +1,177 @@
+      if (image_buffer == NULL)
+        goto error;
+
+      d_blob = new char[image_size];
```
Are you intending to hook `d_blob` into a message output later? Until we use it, I think we can leave it out as it's confusing

reiinakano replied:
Actually, this is a workaround for not being able to use message passing. As soon as the flow graph is closed, all the message passing blocks are also closed. This is the same reason I can't return the <1MB of data from the previous PR.

But for other stuff, like demodulated packets, etc., that are sent *while* the flowgraph is running, I already use enqueue_msg_sink which works fine.

anuraaga replied:
As discussed offline, if this lets all the other blocks run so we get all pending data, it seems like an ok, but sort of hacky, solution. It'd be good to make sure that and make sure there isn't a cleaner way though by going through the APIs once

reiinakano replied:
I opened an issue here: https://github.com/gnuradio/gnuradio/issues/1735 in case the maintainers can help point us in the right direction.

I examined the API again and didn't see anything that could help us. 

One solution I'm considering is this: Currently, we get data from `enqueue_msg_sink` blocks by calling their `observe` method and grabbing a queue, into which they send their messages (which is just raw binary data). Perhaps instead of detecting `enqueue_msg_sink` blocks, we could just detect if a block has an `observe` method (maybe rename to `starcoder_observe` to prevent conflict). This waterfall block can then just drop the final .png into its queue. Technically, it's not really different from what we're planning to do, but at least we won't need to change the Starcoder code. 

The only problem really is that for this special case, we're sending a message while closing the flowgraph. At this point, the message can't be picked up by another block anymore. For any other messages we send *during* runtime (decoded packets, partial.bins), we can just wire it up to an `enqueue_msg_sink` block as designed, and it'll automatically send it out without needing to do anything special.

anuraaga replied:
I was hoping `enqueue_msg_sink` is just a stopgap since polling a queue is reasonably inefficient - we should be able to directly handle messages without polling. I was thinking that at some point, we'd use message-passing to allow directly handling them instead of polling the queue which is why I'm interested in finding a way to make it work.

reiinakano replied:
I've thought about this too but I get stuck when I think about how I can pass a Go callback to a block  since it has to go through Python too. I like Go but I wonder if we overcomplicated things by using Go instead of just Python.

anuraaga replied:
Much of the world is migrating from Python -> Go since the latter provides similar expressiveness with much better performance, so I see starcoder as a way to bring this to gnuradio too. While it's not trivial to interface golang to gnuradio, I think it's worth doing and think it shouldn't be super-hard - all the python just wraps c++ classes, meaning you must be able to access the C++ API directly from golang as well (maybe with SWIG_ConvertPtr or https://github.com/gnuradio/gnuradio/blob/master/gnuradio-runtime/lib/hier_block2.cc#L66) - I don't see any reason it wouldn't work with some swig files and glue code.

But it's ok to think about it while moving along with stopgaps like enqueue_msg_sink, and maybe even this `stop` method, still we should see how we can make the binding more golang-native

anuraaga commented on gr-starcoder/lib/waterfall_heatmap_impl.h:
```diff
@@ -0,0 +1,86 @@
+                    gr_vector_void_star &output_items);
+
+      size_t
+      compute_decimation(int8_t *out, const gr_complex *in, size_t n_fft);
```
Think these helpers can be `private`

anuraaga commented on gr-starcoder/lib/waterfall_heatmap_impl.cc:
```diff
@@ -0,0 +1,298 @@
+/* -*- c++ -*- */
```
Add our license as well to copied files

`
/*
Infostellar license
*/
/*
sagnogs license
*/
`

reiinakano commented:
I've added the tests and addressed comments. I also modified the plotter to add grid labels, etc.

Sample image from one of our passes:
![wut](https://user-images.githubusercontent.com/18363734/38986528-9cb60022-4407-11e8-8327-f45654fc5100.png)

anuraaga commented on gr-starcoder/python/qa_waterfall_plotter.py:
```diff
@@ -0,0 +1,54 @@
+    def setUp (self):
+        self.tb = gr.top_block ()
+        self.fft_size = 40
+        self.filename = "waterfall.png"
```
Can you set this to a `tempfile.mktemp`?

reiinakano replied:
Done
